torch/header_only_apis.txt

```diff
@@ -0,0 +1,49 @@
+# This file contains all the header-only C++ APIs/symbols in torch.
+# If a symbol is added in this file, it should be tested in a .cpp file
+# to guarantee that compiling these symbols do not require linking libtorch
+# to ensure header-only-ness.
+
+# c10/util/TypeCast.h
+convert
+
+# c10/util/bit_cast.h
+bit_cast
+
+# c10/util/BFloat16-math.h, c10/util/BFloat16.h
+BFloat16
+
+# c10/util/Float8_e4m3fn.h
+Float8_e4m3fn
+
+# c10/util/Float8_e4m3fnuz.h
+Float8_e4m3fnuz
+
+# c10/util/Float8_e5m2.h
+Float8_e5m2
+
+# c10/util/Float8_e5m2fnuz.h
+Float8_e5m2fnuz
+
+# c10/util/Half.h
+Half
+
+# c10/util/complex.h
+complex
+
+# ATen/NumericUtils.h, c10/util/generic_math.h
+div_floor_floating
+div_floor_integer
+_isnan
+
+# ATen/core/PhiloxRNGEngine.h
+Philox4_32
+randn
+
+# ATen/cpu/vec/vec.h
+vec::Vectorized
+loadu
+maximum
+size
+minimum
+convert
+clamp_min
```

Comment:
sort?

also this highlights a couple problems:
1) tests that do `using namespace at::vec;` or put their code in `namespace at::vec {` and then test Vectorized will not count
2) any test that checks the size() of any container at all, such as by looping over the indices in a vector, will count
3) method names that commonly occur as substrings of other things (again, such as size -- size_t, max_size) will be prone to false positives
4) common function names (this file itself has convert in it twice!) can't have their individual instances differentiated

Reply:
Yes, will definitely sort! The order for these had just been following the order they appear in the existing tests.

Yea I thought about all of these things on Wednesday and considered building a mapping of symbol name -> regex to look for but that didn't solve all these problems so I went with simple string matching first. I expect using regexes could solve 1 and 3, but to do a good job with everything else I'm thinking we'd need a more complex system. For now I think simple string matching is okay because the workflow I imagine would be:
1. someone wants to move an API to be guaranteed header only
2. they move the API
3. they add the API to the torch/header_only_apis.txt
4. they open a PR
5. i get pinged! and i check that they properly added a cpp test without linking libtorch.so